### PR TITLE
General zipper

### PR DESCRIPTION
**Pull request by andimarek**
There are currently two zipper implementations: for the Ast and for the ExecutionResultNode.

This PR generalises the zipper concept (including the AstTransformer) inside the util package.

### CODE REVIEW

File: src/main/java/graphql/util/NodeAdapter.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Adapts an arbitrary class to behave as a node.
+ * We are using an Adapter because we don't want to require Nodes to implement a certain Interface.
+ *
+ * @param <T>
+ */
+@PublicApi
```
Comment (bbakerman) [inline src/main/java/graphql/util/NodeAdapter.java]: Good thinking!

File: src/main/java/graphql/util/TraverserVisitorStub.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    public TraversalControl enter(TraverserContext<T> context) {
+        return TraversalControl.CONTINUE;
+    }
+
+    @Override
+    public TraversalControl leave(TraverserContext<T> context) {
+        return TraversalControl.CONTINUE;
+    }
+
+}
```
Comment (bbakerman) [inline src/main/java/graphql/util/TraverserVisitorStub.java]: Sometimes I wont if default methods on interfaces is not a better way to do this.  eg some vistor extends TraverserVisitor but its all implemented like the above.  So you only need 1 interface type and not the interface and stub

**Comment by bbakerman** [comment]
The lack of JavaDoc stands out for me.  How would some be an adaptor and what responsibilities do they have?  I know you know - will all people?

File: src/main/java/graphql/util/TraverserVisitorStub.java (hunk 2 above)
Comment (andimarek) [reply src/main/java/graphql/util/TraverserVisitorStub.java]: yea ... don't have an answer to that, maybe it is better do have just default methods.

**Pull request merged**